crates/noirc_evaluator/src/ssa/node.rs

```diff
@@ -0,0 +1,948 @@
+    }
+
+    // Simplifies trivial Phi instructions by returning:
+    // None, if the instruction is unreachable or in the root block and can be safely deleted
+    // Some(id), if the instruction can be replaced by the node id
+    // Some(ins_id), if the instruction is not trivial
+    pub fn simplify_phi(
+        ins_id: arena::Index,
+        phi_arguments: &[(arena::Index, arena::Index)],
+    ) -> Option<arena::Index> {
+        let mut same = None;
+        for op in phi_arguments {
+            if Some(op.0) == same || op.0 == ins_id {
+                continue;
+            }
+            if same.is_some() {
+                //no simplification
+                return Some(ins_id);
+            }
+
+            same = Some(op.0);
+        }
+        //if same.is_none()  => unreachable phi or in root block, can be replaced by ins.lhs (i.e the root) then.
+        same
+    }
+
+    pub fn standard_form(&mut self) {
+        match self.operator {
+            //convert greater into less
+            Operation::ugt => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::ult
+            }
+            Operation::uge => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::ule
+            }
+            Operation::sgt => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::slt
+            }
+            Operation::sge => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::sle
+            }
+            Operation::gt => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::lt
+            }
+            Operation::gte => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::lte
+            }
+            //TODO replace a<b with a<=b+1, but beware of edge cases!
+            Operation::eq_gate => {
+                if self.rhs == self.lhs {
+                    self.rhs = self.idx;
+                    self.is_deleted = true;
+                }
+            }
+            _ => (),
+        }
+        if is_commutative(self.operator) && self.rhs < self.lhs {
+            std::mem::swap(&mut self.rhs, &mut self.lhs);
+        }
+    }
+}
+
+//adpated from LLVM IR
+#[allow(non_camel_case_types)]
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
+pub enum Operation {
+    add,      //(+)
+    safe_add, //(+) safe addtion
+    sub,      //(-)
+    safe_sub, //(-) safe substraction
+    mul,      //(*)
+    safe_mul, //(*) safe multiplication
+    udiv,     //(/) unsigned division
+    sdiv,     //(/) signed division
+    urem,     //(%) modulo; remainder of unsigned division
+    srem,     //(%) remainder of signed division
+    fmod,     //(%) remainder of the floating point division
```

Comment:
Can we remove these floating point operations and open an issue for them? I don't think they will be supported for a while as they are super low priority

Reply:
I was wrong to say they are super low priority as many finance based programs will need them. Maybe just low/medium priority as all of our initial applications like joint-split will not require them?

Reply:
How difficult would supporting floating point types be in noir? If it is so complex to model integers I cannot imagine how complex it'd be to imagine a floating point type. If they are difficult, maybe a Decimal, Fraction, or FixedPoint type would be easier.